File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/query/SpecPath.java
package top.bettercode.summer.data.jpa.query;


import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaBuilder.Trimspec;
import javax.persistence.criteria.Expression;
import javax.persistence.criteria.Path;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

/**
 * @author Peter Wu
 */
public class SpecPath<T, M extends SpecMatcher<T, M>> implements SpecPredicate<T, M> {

  private static final Logger log = LoggerFactory.getLogger(SpecPath.class);

  private final M specMatcher;
  /**
   * name of the attribute
   */
  private final String propertyName;

  private boolean ignoreCase = false;

  private boolean ignoredPath = false;

  private Trimspec trimspec = null;

  private PathMatcher matcher = PathMatcher.EQ;

  private Object value;

  private boolean setValue = false;

  public SpecPath(M specMatcher, String propertyName) {
    this.specMatcher = specMatcher;
    this.propertyName = propertyName;
  }

  //--------------------------------------------

  public static String containing(Object value) {
    return "%" + value + "%";
  }

  public static String ending(Object value) {
    return "%" + value;
  }

  public static String starting(Object value) {
    return value + "%";
  }

  public static <T> Path<?> toPath(Root<T> root, String propertyName) {
    try {
      String[] split = propertyName.split("\\.");
      Path<?> path = null;
      for (String s : split) {
        path = path == null ? root.get(s) : path.get(s);
      }
      return path;
    } catch (IllegalArgumentException e) {
      if (log.isDebugEnabled()) {
        log.debug(e.getMessage());
      }
      return null;
    }
  }
  //--------------------------------------------

  public Path<?> toPath(Root<T> root) {
    return toPath(root, this.propertyName);
  }

  @SuppressWarnings({"rawtypes", "unchecked"})
  @Override
  public Predicate toPredicate(Root<T> root, CriteriaBuilder criteriaBuilder) {
    if (this.isIgnoredPath()) {
      return null;
    }
    Path path = this.toPath(root);
    if (path == null) {
      return null;
    }
    PathMatcher matcher = this.matcher;
    Class pathJavaType = path.getJavaType();
    switch (matcher) {
      case IS_TRUE:
        return criteriaBuilder.isTrue(path);
      case IS_FALSE:
        return criteriaBuilder.isFalse(path);
      case IS_NULL:
        return criteriaBuilder.isNull(path);
      case IS_NOT_NULL:
        return criteriaBuilder.isNotNull(path);
    }
    if (pathJavaType.equals(String.class)) {
      switch (matcher) {
        case IS_EMPTY:
          return criteriaBuilder.equal(path, "");
        case IS_NOT_EMPTY:
          return criteriaBuilder.notEqual(path, "");
        case IS_NULL_OR_EMPTY:
          return criteriaBuilder.or(criteriaBuilder.isNull(path),
              criteriaBuilder.equal(path, ""));
        case IS_NOT_NULL_OR_EMPTY:
          return criteriaBuilder.and(criteriaBuilder.isNotNull(path),
              criteriaBuilder.notEqual(path, ""));
      }
    }
    if (!setValue) {
      return null;
    }
    Object value = this.value;
    if (value != null) {
      switch (matcher) {
        case BETWEEN:
          Assert.isTrue(value instanceof SpecPath.BetweenValue,
              "BETWEEN matcher with wrong value");
          BetweenValue betweenValue = (BetweenValue) value;
          return criteriaBuilder.between(path, betweenValue.getFirst(),
              betweenValue.getSecond());
        case GT:
          return criteriaBuilder.greaterThan(path, (Comparable) value);
        case GE:
          return criteriaBuilder.greaterThanOrEqualTo(path, (Comparable) value);
        case LT:
          return criteriaBuilder.lessThan(path, (Comparable) value);
        case LE:
          return criteriaBuilder.lessThanOrEqualTo(path, (Comparable) value);
      }
    }

    if (pathJavaType.equals(String.class)) {
      Expression<String> stringExpression = path;
      boolean ignoreCase = this.ignoreCase;
      if (ignoreCase) {
        stringExpression = criteriaBuilder.lower(stringExpression);
        if (value instanceof String) {
          value = value.toString().toLowerCase();
        }
      }
      if (trimspec != null && value instanceof String) {
        switch (trimspec) {
          case LEADING:
            value = StringUtils.trimLeadingWhitespace((String) value);
            break;
          case TRAILING:
            value = StringUtils.trimTrailingWhitespace((String) value);
            break;
          case BOTH:
            value = StringUtils.trimWhitespace((String) value);
            break;
        }
      }

      switch (matcher) {
        case EQ:
          return criteriaBuilder.equal(stringExpression, value);
        case NE:
          return criteriaBuilder.notEqual(stringExpression, value);
        case IN:
          Assert.isTrue(value instanceof Collection, "IN matcher with wrong value");
          List<String> collect = ((Collection<?>) value).stream()
              .map(s -> ignoreCase ? s.toString().toLowerCase() : s.toString())
              .collect(Collectors.toList());
          return stringExpression.in(collect);
        case NOT_IN:
          Assert.isTrue(value instanceof Collection, "IN matcher with wrong value");
          List<String> notInCollect = ((Collection<?>) value).stream()
              .map(s -> ignoreCase ? s.toString().toLowerCase() : s.toString())
              .collect(Collectors.toList());
          return criteriaBuilder.not(stringExpression.in(notInCollect));
      }
      if (value != null) {
        switch (matcher) {
          case LIKE:
            return criteriaBuilder.like(stringExpression, (String) value);
          case NOT_LIKE:
            return criteriaBuilder.notLike(stringExpression, (String) value);
          case STARTING:
            return criteriaBuilder.like(stringExpression, starting(value));
          case NOT_STARTING:
            return criteriaBuilder.notLike(stringExpression, starting(value));
          case ENDING:
            return criteriaBuilder.like(stringExpression, ending(value));
          case NOT_ENDING:
            return criteriaBuilder.notLike(stringExpression, ending(value));
          case CONTAINING:
            return criteriaBuilder.like(stringExpression, containing(value));
          case NOT_CONTAINING:
            return criteriaBuilder.notLike(stringExpression, containing(value));
        }
      }
    } else {
      switch (matcher) {
        case EQ:
          return criteriaBuilder.equal(path, value);
        case NE:
          return criteriaBuilder.notEqual(path, value);
        case IN:
          Assert.isTrue(value instanceof Collection, "IN matcher with wrong value");
          Collection<?> collect = ((Collection<?>) value);
          return path.in(collect);
        case NOT_IN:
          Assert.isTrue(value instanceof Collection, "IN matcher with wrong value");
          Collection<?> notInCollect = ((Collection<?>) value);
          return criteriaBuilder.not(path.in(notInCollect));
      }
    }
    return null;
  }

  //--------------------------------------------


  public boolean isSetValue() {
    return setValue;
  }

  public boolean isIgnoredPath() {
    return this.ignoredPath;
  }

  public SpecPath<T, M> setValue(Object value) {
    this.value = value;
    this.setValue = true;
    return this;
  }

  public M sortBy(Direction direction) {
    this.specMatcher.sortBy(direction, this.propertyName);
    return this.specMatcher;
  }

  public M asc() {
    return this.sortBy(Direction.ASC);
  }

  public M desc() {
    return this.sortBy(Direction.DESC);
  }

  public M ignoreCase() {
    ignoreCase = true;
    return this.specMatcher;
  }

  public M trim(Trimspec trimspec) {
    this.trimspec = trimspec;
    return this.specMatcher;
  }

  public M trim() {
    return this.trim(Trimspec.BOTH);
  }

  public M ignoredPath() {
    ignoredPath = true;
    return this.specMatcher;
  }

  public M withMatcher(PathMatcher matcher) {
    this.matcher = matcher;
    return this.specMatcher;
  }

  public M withMatcher(Object value, PathMatcher matcher) {
    this.setValue(value);
    return this.withMatcher(matcher);
  }

  public M isTrue() {
    return this.withMatcher(PathMatcher.IS_TRUE);
  }

  public M isFalse() {
    return this.withMatcher(PathMatcher.IS_FALSE);
  }

  public M isNull() {
    return this.withMatcher(PathMatcher.IS_NULL);
  }

  public M isNotNull() {
    return this.withMatcher(PathMatcher.IS_NOT_NULL);
  }

  public M isEmpty() {
    return this.withMatcher(PathMatcher.IS_EMPTY);
  }

  public M isNotEmpty() {
    return this.withMatcher(PathMatcher.IS_NOT_EMPTY);
  }

  public M isNullOrEmpty() {
    return this.withMatcher(PathMatcher.IS_NULL_OR_EMPTY);
  }

  public M isNotNullOrEmpty() {
    return this.withMatcher(PathMatcher.IS_NOT_NULL_OR_EMPTY);
  }

  public M equal() {
    return this.withMatcher(PathMatcher.EQ);
  }

  public M notEqual() {
    return this.withMatcher(PathMatcher.NE);
  }

  public M gt() {
    return this.withMatcher(PathMatcher.GT);
  }

  public M ge() {
    return this.withMatcher(PathMatcher.GE);
  }

  public M lt() {
    return this.withMatcher(PathMatcher.LT);
  }

  public M le() {
    return this.withMatcher(PathMatcher.LE);
  }

  public M like() {
    return this.withMatcher(PathMatcher.LIKE);
  }

  public M starting() {
    return this.withMatcher(PathMatcher.STARTING);
  }

  public M ending() {
    return this.withMatcher(PathMatcher.ENDING);
  }

  public M containing() {
    return this.withMatcher(PathMatcher.CONTAINING);
  }

  public M notStarting() {
    return this.withMatcher(PathMatcher.NOT_STARTING);
  }

  public M notEnding() {
    return this.withMatcher(PathMatcher.NOT_ENDING);
  }

  public M notContaining() {
    return this.withMatcher(PathMatcher.NOT_CONTAINING);
  }

  public M notLike() {
    return this.withMatcher(PathMatcher.NOT_LIKE);
  }

  //--------------------------------------------
  public M equal(Object value) {
    return withMatcher(value, PathMatcher.EQ);
  }

  public M notEqual(Object value) {
    return withMatcher(value, PathMatcher.NE);
  }

  public <Y extends Comparable<? super Y>> M gt(Y value) {
    return withMatcher(value, PathMatcher.GT);
  }

  public <Y extends Comparable<? super Y>> M ge(Y value) {
    return withMatcher(value, PathMatcher.GE);
  }

  public <Y extends Comparable<? super Y>> M lt(Y value) {
    return withMatcher(value, PathMatcher.LT);
  }

  public <Y extends Comparable<? super Y>> M le(Y value) {
    return withMatcher(value, PathMatcher.LE);
  }

  public <Y extends Comparable<? super Y>> M between(Y first, Y second) {
    return withMatcher(new BetweenValue<>(first, second), PathMatcher.BETWEEN);
  }

  public M like(String value) {
    return withMatcher(value, PathMatcher.LIKE);
  }

  public M starting(String value) {
    return withMatcher(value, PathMatcher.STARTING);
  }

  public M ending(String value) {
    return withMatcher(value, PathMatcher.ENDING);
  }

  public M containing(String value) {
    return withMatcher(value, PathMatcher.CONTAINING);
  }

  public M notStarting(String value) {
    return withMatcher(value, PathMatcher.NOT_STARTING);
  }

  public M notEnding(String value) {
    return withMatcher(value, PathMatcher.NOT_ENDING);
  }

  public M notContaining(String value) {
    return withMatcher(value, PathMatcher.NOT_CONTAINING);
  }

  public M notLike(String value) {
    return withMatcher(value, PathMatcher.NOT_LIKE);
  }

  @SafeVarargs
  public final <E> M in(E... value) {
    return withMatcher(Arrays.asList(value), PathMatcher.IN);
  }

  public M in(Collection<?> value) {
    return withMatcher(value, PathMatcher.IN);
  }

  @SafeVarargs
  public final <E> M notIn(E... value) {
    return withMatcher(Arrays.asList(value), PathMatcher.NOT_IN);
  }

  public M notIn(Collection<?> value) {
    return withMatcher(value, PathMatcher.NOT_IN);
  }

  static class BetweenValue<Y extends Comparable<? super Y>> {

    private final Y first;
    private final Y second;

    public BetweenValue(Y first, Y second) {
      this.first = first;
      this.second = second;
    }

    public Y getFirst() {
      return first;
    }

    public Y getSecond() {
      return second;
    }
  }
}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/query/SpecPath.kt
package top.bettercode.summer.data.jpa.query

import org.slf4j.LoggerFactory
import org.springframework.data.domain.Sort
import org.springframework.util.Assert
import java.util.*
import javax.persistence.criteria.*
import javax.persistence.criteria.CriteriaBuilder.Trimspec

/**
 * @author Peter Wu
 */
class SpecPath<P, T : Any?, M : SpecMatcher<T, M>>(
    private val specMatcher: M,
    /**
     * name of the attribute
     */
    val propertyName: String
) : SpecPredicate<T, M> {
    /**
     * 忽略大小写
     */
    private var ignoreCase = false

    /**
     * 不参与计算查询条件
     */
    var isIgnored = false
        private set

    /**
     * 用于指定如何修剪字符串。
     */
    private var trimspec: Trimspec? = null

    /**
     * 条件匹配方式
     */
    private var matcher = PathMatcher.EQ

    /**
     * 条件查询值
     */
    var criteria: P? = null
        private set

    /**
     * 条件查询值2,用于范围查询
     */
    var criteria2: P? = null
        private set

    /**
     * 集合条件查询值,用于范围查询
     */
    var criterias: Collection<P>? = null
        private set

    /**
     * 更新值
     */
    var criteriaUpdate: P? = null
        private set

    //--------------------------------------------
    var isSetCriteria: Boolean = false

    var isSetCriteriaUpdate: Boolean = false

    //--------------------------------------------
    fun toPath(root: Root<T>): Path<*> {
        return toPath(root, propertyName)
    }

    @Suppress("UNCHECKED_CAST")
    override fun toPredicate(root: Root<T>, criteriaBuilder: CriteriaBuilder): Predicate? {
        if (isIgnored) {
            return null
        }
        val path = this.toPath(root)
        val matcher = matcher
        val pathJavaType = path.javaType
        when (matcher) {
            PathMatcher.IS_TRUE -> return criteriaBuilder.isTrue(path as Expression<Boolean>)
            PathMatcher.IS_FALSE -> return criteriaBuilder.isFalse(path as Expression<Boolean>)
            PathMatcher.IS_NULL -> return criteriaBuilder.isNull(path)
            PathMatcher.IS_NOT_NULL -> return criteriaBuilder.isNotNull(path)
            else -> {}
        }
        if (pathJavaType == String::class.java) {
            when (matcher) {
                PathMatcher.IS_EMPTY -> return criteriaBuilder.equal(path, "")
                PathMatcher.IS_NOT_EMPTY -> return criteriaBuilder.notEqual(path, "")
                PathMatcher.IS_NULL_OR_EMPTY -> return criteriaBuilder.or(
                    criteriaBuilder.isNull(path),
                    criteriaBuilder.equal(path, "")
                )

                PathMatcher.IS_NOT_NULL_OR_EMPTY -> return criteriaBuilder.and(
                    criteriaBuilder.isNotNull(path),
                    criteriaBuilder.notEqual(path, "")
                )

                else -> {}
            }
        }
        if (!isSetCriteria) {
            return null
        }
        if (criteria != null) {
            when (matcher) {
                PathMatcher.BETWEEN -> {
                    return criteriaBuilder.between(
                        path as Expression<Comparable<Comparable<*>>>,
                        criteria as Comparable<Comparable<*>>,
                        criteria2 as Comparable<Comparable<*>>
                    )
                }

                PathMatcher.GT -> return criteriaBuilder.greaterThan(
                    path as Expression<Comparable<Comparable<*>>>,
                    criteria as Comparable<Comparable<*>>
                )

                PathMatcher.GE -> return criteriaBuilder.greaterThanOrEqualTo(
                    path as Expression<Comparable<Comparable<*>>>,
                    criteria as Comparable<Comparable<*>>
                )

                PathMatcher.LT -> return criteriaBuilder.lessThan(
                    path as Expression<Comparable<Comparable<*>>>,
                    criteria as Comparable<Comparable<*>>
                )

                PathMatcher.LE -> return criteriaBuilder.lessThanOrEqualTo(
                    path as Expression<Comparable<Comparable<*>>>,
                    criteria as Comparable<Comparable<*>>
                )

                else -> {}
            }
        }
        if (pathJavaType == String::class.java) {
            var criteria = this.criteria
            var stringExpression: Expression<String> = path as Expression<String>
            if (criteria != null) {
                val ignoreCase = ignoreCase
                if (ignoreCase) {
                    stringExpression = criteriaBuilder.lower(stringExpression)
                    if (criteria is String) {
                        criteria = criteria.lowercase(Locale.getDefault()) as P
                    }
                }
                if (trimspec != null && criteria is String) {
                    when (trimspec) {
                        Trimspec.LEADING -> criteria = criteria.trimStart() as P
                        Trimspec.TRAILING -> criteria = criteria.trimEnd() as P
                        Trimspec.BOTH -> criteria = criteria.trim() as P
                        else -> {}
                    }
                }
                criteria as Any

                when (matcher) {
                    PathMatcher.LIKE -> return criteriaBuilder.like(
                        stringExpression,
                        criteria as String
                    )

                    PathMatcher.NOT_LIKE -> return criteriaBuilder.notLike(
                        stringExpression,
                        criteria as String
                    )

                    PathMatcher.STARTING -> return criteriaBuilder.like(
                        stringExpression,
                        criteria.starting()
                    )

                    PathMatcher.NOT_STARTING -> return criteriaBuilder.notLike(
                        stringExpression,
                        criteria.starting()
                    )

                    PathMatcher.ENDING -> return criteriaBuilder.like(
                        stringExpression,
                        criteria.ending()
                    )

                    PathMatcher.NOT_ENDING -> return criteriaBuilder.notLike(
                        stringExpression,
                        criteria.ending()
                    )

                    PathMatcher.CONTAINING -> return criteriaBuilder.like(
                        stringExpression,
                        criteria.containing()
                    )

                    PathMatcher.NOT_CONTAINING -> return criteriaBuilder.notLike(
                        stringExpression,
                        criteria.containing()
                    )

                    else -> {}
                }
            }
            when (matcher) {
                PathMatcher.EQ -> return criteriaBuilder.equal(stringExpression, criteria)
                PathMatcher.NE -> return criteriaBuilder.notEqual(stringExpression, criteria)
                PathMatcher.IN -> {
                    Assert.notNull(criterias, "IN matcher with wrong criteria")
                    val collect = criterias!!
                        .map { s: P ->
                            if (ignoreCase) s.toString()
                                .lowercase(Locale.getDefault()) else s.toString()
                        }
                    return stringExpression.`in`(collect)
                }

                PathMatcher.NOT_IN -> {
                    Assert.notNull(criterias, "IN matcher with wrong criteria")
                    val notInCollect = criterias!!
                        .map { s: P ->
                            if (ignoreCase) s.toString()
                                .lowercase(Locale.getDefault()) else s.toString()
                        }
                    return criteriaBuilder.not(stringExpression.`in`(notInCollect))
                }

                else -> {}
            }
        } else {
            when (matcher) {
                PathMatcher.EQ -> return criteriaBuilder.equal(path, criteria)
                PathMatcher.NE -> return criteriaBuilder.notEqual(path, criteria)
                PathMatcher.IN -> {
                    Assert.notNull(criterias, "IN matcher with wrong criteria")
                    return path.`in`(criterias)
                }

                PathMatcher.NOT_IN -> {
                    Assert.notNull(criterias, "IN matcher with wrong criteria")
                    return criteriaBuilder.not(path.`in`(criterias))
                }

                else -> {}
            }
        }
        return null
    }

    fun criteria(criteria: P?): SpecPath<P, T, M> {
        this.isSetCriteria = true
        this.criteria = criteria
        return this
    }

    fun criteriaUpdate(criteriaUpdate: P?): M {
        this.isSetCriteriaUpdate = true
        this.criteriaUpdate = criteriaUpdate
        return specMatcher
    }

    fun sortBy(direction: Sort.Direction): M {
        specMatcher.sortBy(direction, propertyName)
        return specMatcher
    }

    fun asc(): M {
        return this.sortBy(Sort.Direction.ASC)
    }

    fun desc(): M {
        return this.sortBy(Sort.Direction.DESC)
    }

    fun ignoreCase(): M {
        ignoreCase = true
        return specMatcher
    }

    @JvmOverloads
    fun trim(trimspec: Trimspec = Trimspec.BOTH): M {
        this.trimspec = trimspec
        return specMatcher
    }

    fun ignored(): M {
        isIgnored = true
        return specMatcher
    }

    fun withMatcher(matcher: PathMatcher): M {
        this.matcher = matcher
        return specMatcher
    }

    fun withMatcher(criteria: P?, matcher: PathMatcher): M {
        this.criteria(criteria)
        return this.withMatcher(matcher)
    }

    val isTrue: M
        get() = this.withMatcher(PathMatcher.IS_TRUE)
    val isFalse: M
        get() = this.withMatcher(PathMatcher.IS_FALSE)
    val isNull: M
        get() = this.withMatcher(PathMatcher.IS_NULL)
    val isNotNull: M
        get() = this.withMatcher(PathMatcher.IS_NOT_NULL)
    val isEmpty: M
        get() = this.withMatcher(PathMatcher.IS_EMPTY)
    val isNotEmpty: M
        get() = this.withMatcher(PathMatcher.IS_NOT_EMPTY)
    val isNullOrEmpty: M
        get() = this.withMatcher(PathMatcher.IS_NULL_OR_EMPTY)
    val isNotNullOrEmpty: M
        get() = this.withMatcher(PathMatcher.IS_NOT_NULL_OR_EMPTY)

    fun equal(): M {
        return eq()
    }

    fun eq(): M {
        return this.withMatcher(PathMatcher.EQ)
    }

    fun notEqual(): M {
        return ne()
    }

    fun ne(): M {
        return this.withMatcher(PathMatcher.NE)
    }

    fun gt(): M {
        return this.withMatcher(PathMatcher.GT)
    }

    fun ge(): M {
        return this.withMatcher(PathMatcher.GE)
    }

    fun lt(): M {
        return this.withMatcher(PathMatcher.LT)
    }

    fun le(): M {
        return this.withMatcher(PathMatcher.LE)
    }

    fun like(): M {
        return this.withMatcher(PathMatcher.LIKE)
    }

    fun starting(): M {
        return this.withMatcher(PathMatcher.STARTING)
    }

    fun ending(): M {
        return this.withMatcher(PathMatcher.ENDING)
    }

    fun containing(): M {
        return this.withMatcher(PathMatcher.CONTAINING)
    }

    fun notStarting(): M {
        return this.withMatcher(PathMatcher.NOT_STARTING)
    }

    fun notEnding(): M {
        return this.withMatcher(PathMatcher.NOT_ENDING)
    }

    fun notContaining(): M {
        return this.withMatcher(PathMatcher.NOT_CONTAINING)
    }

    fun notLike(): M {
        return this.withMatcher(PathMatcher.NOT_LIKE)
    }

    //--------------------------------------------

    fun equal(criteria: P?): M {
        return eq(criteria)
    }

    fun eq(criteria: P?): M {
        return withMatcher(criteria, PathMatcher.EQ)
    }

    fun notEqual(criteria: P?): M {
        return ne(criteria)
    }

    fun ne(criteria: P?): M {
        return withMatcher(criteria, PathMatcher.NE)
    }

    fun gt(criteria: P): M {
        return withMatcher(criteria, PathMatcher.GT)
    }

    fun ge(criteria: P): M {
        return withMatcher(criteria, PathMatcher.GE)
    }

    fun lt(criteria: P): M {
        return withMatcher(criteria, PathMatcher.LT)
    }

    fun le(criteria: P): M {
        return withMatcher(criteria, PathMatcher.LE)
    }

    fun between(first: P, second: P): M {
        this.criteria2 = second
        return withMatcher(first, PathMatcher.BETWEEN)
    }

    @Suppress("UNCHECKED_CAST")
    fun like(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.LIKE)
    }

    @Suppress("UNCHECKED_CAST")
    fun starting(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.STARTING)
    }

    @Suppress("UNCHECKED_CAST")
    fun ending(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.ENDING)
    }

    @Suppress("UNCHECKED_CAST")
    fun containing(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.CONTAINING)
    }

    @Suppress("UNCHECKED_CAST")
    fun notStarting(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.NOT_STARTING)
    }

    @Suppress("UNCHECKED_CAST")
    fun notEnding(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.NOT_ENDING)
    }

    @Suppress("UNCHECKED_CAST")
    fun notContaining(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.NOT_CONTAINING)
    }

    @Suppress("UNCHECKED_CAST")
    fun notLike(criteria: String?): M {
        return withMatcher(criteria as P?, PathMatcher.NOT_LIKE)
    }

    fun `in`(vararg criteria: P): M {
        this.isSetCriteria = true
        this.criterias = criteria.toList()
        return this.withMatcher(PathMatcher.IN)
    }

    fun `in`(criterias: Collection<P>): M {
        this.isSetCriteria = true
        this.criterias = criterias
        return this.withMatcher(PathMatcher.IN)
    }

    fun notIn(vararg criteria: P): M {
        this.isSetCriteria = true
        this.criterias = criteria.toList()
        return this.withMatcher(PathMatcher.NOT_IN)
    }

    fun notIn(criterias: Collection<P>): M {
        this.isSetCriteria = true
        this.criterias = criterias
        return this.withMatcher(PathMatcher.NOT_IN)
    }

    companion object {
        private val log = LoggerFactory.getLogger(SpecPath::class.java)

        //--------------------------------------------
        fun Any.containing(): String {
            return "%$this%"
        }

        fun Any.ending(): String {
            return "%$this"
        }

        fun Any.starting(): String {
            return "$this%"
        }

        fun <T> toPath(root: Root<T>, propertyName: String): Path<*> {
            val split = propertyName.split(".")
            var path: Path<*>? = null
            for (s in split) {
                path = if (path == null) root.get<Any>(s) else path.get<Any>(s)
            }
            return path ?: throw IllegalArgumentException("Path not found: $propertyName")
        }
    }
}
